### PR TITLE
Feature/unsup multichan waveform dataset

### PR DESCRIPTION
**Pull request by kkarrancsu**
Added Dataset which supports multichannel audio-samples.  Updated collator to drop pad-tracks.

### CODE REVIEW

File: lhotse/dataset/collation.py [hunk 1]
```diff
@@ -315,6 +315,44 @@ def collate_multi_channel_audio(cuts: CutSet) -> torch.Tensor:
...
+    The output shape is ``(batch, channel, time)``.
+    The cuts will be padded with silence if necessary.
+    """
+    assert all(cut.has_recording for cut in cuts)
+    assert all(isinstance(cut, MixedCut) for cut in cuts)
+    
+    # TODO: how to ensure that each track is synced across batches?  i.e. dim=1 is the track index
+    #  and should correspond to the same mic across batches 
+
+    cuts = maybe_pad(cuts)
```
Comment (pzelasko) [inline lhotse/dataset/collation.py:339]: This should not be needed as you're manually zero-padding later.

File: lhotse/dataset/collation.py [hunk 2]
```diff
@@ -315,6 +315,44 @@ def collate_multi_channel_audio(cuts: CutSet) -> torch.Tensor:
...
+    cuts = remove_pad_tracks(cuts)
```
Comment (pzelasko) [inline lhotse/dataset/collation.py:340]: I think there is a pitfall here, what if a MixedCut looks like:

```
|-------cut1-------||---padding---||----cut2----|
```

or any variation of the situation where padding is in between of two cuts. I don't think Lhotse would handle these situations well with your current code. Maybe you should try only removing the padding at the end (and beginning, but for that one you have to be careful about modifying the offsets on the remaining tracks). Rather than manually removing PaddingCuts, I suggest using `.truncate()` with carefully computed offset and duration arguments; that method will handle a lot of pitfalls and edge-cases.

File: lhotse/dataset/collation.py [hunk 3]
```diff
@@ -315,6 +315,44 @@ def collate_multi_channel_audio(cuts: CutSet) -> torch.Tensor:
...
+    #  this is zero-padding but that seems bad ...
+    max_ntrack = max(len(cut.tracks) for cut in cuts)
+    max_nsamp = max(cut.num_samples for cut in cuts)
+
+    # NOTE: this ends up zero-padding! is that appropriate?
+    audio = torch.zeros(len(cuts), max_ntrack, max_nsamp) 
+    for idx, cut in enumerate(cuts):
+        ntrack = len(cut.tracks)
+        nsamp = cut.num_samples
+        audio[idx, 0:ntrack, 0:nsamp] = torch.from_numpy(cut.load_audio(mixed=False))
```
Comment (pzelasko) [inline lhotse/dataset/collation.py:352]: Note that if you did `cut.mix(musan_cut)` here, it will also add an extra track; as is, the code would not work with additive noise data augmentation

File: lhotse/dataset/collation.py [hunk 4]
```diff
@@ -315,6 +315,44 @@ def collate_multi_channel_audio(cuts: CutSet) -> torch.Tensor:
...
+    assert all(isinstance(cut, MixedCut) for cut in cuts)
+    
+    # TODO: how to ensure that each track is synced across batches?  i.e. dim=1 is the track index
+    #  and should correspond to the same mic across batches 
+
+    cuts = maybe_pad(cuts)
+    cuts = remove_pad_tracks(cuts)
+    
+    # NOTE: what to do when the # of tracks is not the same across cuts, right now
+    #  this is zero-padding but that seems bad ...
```
Comment (pzelasko) [inline lhotse/dataset/collation.py:343]: I think you won't escape zero-padding of examples with less channels if you need to collate the data. However, I suggest you modify this function to return a 3-tuple: `(audio, audio_lens, channel_indexes)` where `audio` is the collated data with shape `(B, C, T)`, `audio_lens` has the length of each multi-channel example of shape `(B,)`, and `channel_indexes` is a list of lists of which `C` dim indexes have meaningful channels for examples (it could also be `channel_lens` tensor of shape `(B,)` assuming first `c` channels are always meaningful, if it's possible to guarantee).

Comment (pzelasko) [reply lhotse/dataset/collation.py:343]: But in the end your models will have to somehow deal with the non-meaningful channels anyway. As long as you're working on same-number-of-channels data, no need to overthink this.

File: lhotse/dataset/collation.py [hunk 5]
```diff
@@ -315,6 +315,44 @@ def collate_multi_channel_audio(cuts: CutSet) -> torch.Tensor:
...
+    Load audio samples for all the cuts and return them as a batch in a torch tensor.
+    The cuts have to be of type ``MixedCut`` and their tracks will be interpreted as individual channels.
+    The output shape is ``(batch, channel, time)``.
+    The cuts will be padded with silence if necessary.
+    """
+    assert all(cut.has_recording for cut in cuts)
+    assert all(isinstance(cut, MixedCut) for cut in cuts)
+    
+    # TODO: how to ensure that each track is synced across batches?  i.e. dim=1 is the track index
+    #  and should correspond to the same mic across batches
```
Comment (pzelasko) [inline lhotse/dataset/collation.py:337]: You can ensure the tracks are sorted by some property; I imagine this is something very corpus specific and should be done by the user, not by the library.

File: lhotse/dataset/unsupervised.py [hunk 6]
```diff
@@ -74,6 +74,41 @@ def _validate(self, cuts: CutSet) -> None:
         assert all(cut.has_recording for cut in cuts)
 
 
+class UnsupervisedMultiChanWaveformDataset(UnsupervisedDataset):
```
Comment (pzelasko) [inline lhotse/dataset/unsupervised.py:77]:
```suggestion
class MultiChannelWaveformDataset(UnsupervisedDataset):
```

somehow reads better to me

File: lhotse/dataset/unsupervised.py [hunk 7]
```diff
@@ -74,6 +74,41 @@ def _validate(self, cuts: CutSet) -> None:
...
+        if self.collate:
+            audio = collate_multi_channel_audio_rmpad(cuts)
+            audio_lens = 0 # TODO
+            return {
+                "cuts": cuts,
+                "audio": audio,
+                "audio_lens": audio_lens,
+            }
+        else:
+            return {"cuts": cuts, "audio": [c.load_audio(mixed=False) for c in cuts]}
```
Comment (pzelasko) [inline lhotse/dataset/unsupervised.py:104]: This line would again have the extra padding channels problem. This suggests that maybe the solution should not be (entirely) in the `collate` function, but inside `load_audio`, e.g. controlled by an extra argument?

File: lhotse/dataset/speech_recognition.py [hunk 8]
```diff
@@ -96,7 +96,7 @@ def __getitem__(self, cuts: CutSet) -> Dict[str, Union[torch.Tensor, List[str]]]
         Return a new batch, with the batch size automatically determined using the constraints
         of max_frames and max_cuts.
         """
-        validate_for_asr(cuts)
+        #validate_for_asr(cuts)
```
Comment (pzelasko) [inline lhotse/dataset/speech_recognition.py]: This should be uncommented

**Review by pzelasko — COMMENTED** [review]
Thanks! I think it's a good start, will take a bit more effort to handle multi-channel properly here, but we'll get there.

**Comment by pzelasko** [comment]
For anybody interested in this, here's some context of our earlier discussion with @kkarrancsu 

I expect it you to run into issues related to padding and MUSAN data augmentation with it. Basically, padding and augmentation creates extra tracks in MixedCut, and neither MixedCut nor collate_multi_channel_audio know which tracks are the data, and which tracks are the padding / noise. So, for 4-channel audio, you might end up with 6-channel output from `collate_multi_channel_audio`; you might want to modify it somehow to avoid that (interpret padding as padding all channels and noise augmentation as… adding the noise to each channel? or just one of them? I don’t know)

**Comment by pzelasko** [comment]
@kkarrancsu I have a different idea -- we could add a new attribute to `MixTrack` that's called `separate_channel: bool = True` -- it would indicate if a given track is a "data" channel or an "augmentation" channel. We would add a parameter with the same name to `mix` (default=True) and `pad` (default=False) operations on all cuts. 

We would need to extend `FeatureMixer` and `AudioMixer` to handle these cases properly. The idea is to again add `separate_channel` parameter to `add_to_mix` method, and modify the  property `unmixed_audio`:

https://github.com/lhotse-speech/lhotse/blob/b41e4f8c57954b94790f2d05de310a11071512bc/lhotse/audio.py#L984-L990

so that instead of simply vstacking the right channels, it vstacks only the "separate" channels, downmixes the remaining channels to mono, and adds them to each of the "separate" channels. The analogous operation is needed for `FeatureMixer`.

Then, `collate_multi_channel_audio` could remain almost unmodified w.r.t. to what's there in the codebase, and it would not need any special logic to figure out what's a real channel and what is just padding/augmentation.

Of course we'd need to add more unit tests to make sure this doesn't break anything and works as expected.

**Comment by danpovey** [comment]
It seems to me that the more "correct" way to do this would be, when adding noise to multi-channel audio, to add multiple channels of noise.  I assume this would require some nontrivial simulation, possibly with multiple sources.
I would have thought that for any application that was going to process multiple-channel audio in a non-trivial way, just adding a single type of noise to all channels would not really be sufficient.

**Comment by danpovey** [comment]
OK, sure, it was just a thought.
